conslogging/bundle.go

```diff
@@ -0,0 +1,246 @@
+		builder.result = result
+	}
+}
+
+func (bb *BundleBuilder) PrefixStatus(prefix, status string) {
+	if builder, ok := bb.logmap[prefix]; ok {
+		builder.status = status
+	}
+}
+
+func (bb *BundleBuilder) PrefixWriter(prefix string) io.Writer {
+	if builder, ok := bb.logmap[prefix]; ok {
+		return builder
+	}
+
+	writer := &TargetLogger{
+		writer:  &strings.Builder{},
+		status:  StatusWaiting,
+		result:  "",
+		started: time.Now(),
+	}
+	bb.logmap[prefix] = writer
+	return writer
+}
+
+func (bb *BundleBuilder) WriteToDisk() error {
+	fmt.Println(bb.RootPath)
+
+	targetPath := path.Join(bb.RootPath, "target")
+	err := os.MkdirAll(targetPath, 0700)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to write targets directory for bundle at %s", bb.RootPath)
+	}
+
+	manifest := &Manifest{
+		Version:    1,
+		Duration:   int(time.Since(bb.started).Seconds()),
+		Status:     "complete",
+		Result:     "success",
+		CreatedAt:  time.Now().In(time.UTC),
+		Entrypoint: bb.Entrypoint,
+		Targets:    make([]TargetManifest, 0),
+	}
+
+	for prefix, lines := range bb.logmap {
+		if lines.writer.Len() <= 0 {
+			// Do not write empty logs, if the prefix didn't write anything
+			continue
+		}
+
+		trimmed := strings.TrimSpace(prefix)
+		escaped := url.PathEscape(trimmed)
+		logPath := path.Join(targetPath, escaped)
+
+		command, summary := bb.GetCommandAndSummary(prefix, lines.writer)
+
+		manifest.Targets = append(manifest.Targets, TargetManifest{
+			Name:     prefix,
+			Status:   lines.status,
+			Result:   lines.result,
+			Duration: int(time.Since(lines.started).Seconds()),
+			Size:     lines.writer.Len(),
+			Command:  command,
+			Summary:  summary,
+		})
+
+		if lines.result != ResultSuccess {
+			manifest.Result = lines.result
+		}
+
+		if lines.status != StatusComplete {
+			manifest.Status = lines.status
+		}
+
+		tgtErr := ioutil.WriteFile(logPath, []byte(lines.writer.String()), 0600)
+		if err != nil {
+			err = multierror.Append(err, tgtErr)
+		}
+	}
+	if err != nil {
+		return errors.Wrap(err, "errors while writing targets for log bundle")
+	}
+
+	manifestJSON, _ := json.Marshal(&manifest)
+	err = ioutil.WriteFile(path.Join(bb.RootPath, "manifest"), manifestJSON, 0600)
+	if err != nil {
+		return errors.Wrap(err, "failed to serialize bundle manifest")
+	}
+
+	permissionsJSON, _ := json.Marshal(&Permissions{
+		Version: 1,
+		Users:   make([]uint64, 0),
+		Orgs:    make([]uint64, 0),
+	})
+	err = ioutil.WriteFile(path.Join(bb.RootPath, "permissions"), permissionsJSON, 0600)
+	if err != nil {
+		return errors.Wrap(err, "failed to serialize bundle permissions placeholder")
+	}
+
+	return nil
+}
+
+// Nobody expects ANSI in the command/summary.
+// So, even if we don't inject color we should strip it since a tool inside could have done an ANSI too. *SIGH*
+const ansi = "[\u001B\u009B][[\\]()#;?]*(?:(?:(?:[a-zA-Z\\d]*(?:;[a-zA-Z\\d]*)*)?\u0007)|(?:(?:\\d{1,4}(?:;\\d{0,4})*)?[\\dA-PRZcf-ntqry=><~]))"
```

Comment:
I am sorry for this and all the other regex. Willing to sign documents saying they are my problem now.